### ISSUE TITLE
How can client detect server disconnection faster?

### ISSUE THREAD

**Issue by Eternal21** [aspnet/AspNetCore#6998]
I'm upgrading a project from SignalR to SignalR Core, and ran into the following issue.  I don't see a way for the .NET client to quickly detect that it has been disconnected from the server.  This happens when the web service recycles its application pool.  It takes 90 seconds for the .NET Core client to detect that, via `HubConnection.Closed` event.

In the original SignalR, this was instantaneous via a number of events like `HubConnection.ConnectionSlow`, `HubConnection.Reconnecting` or `HubConnection.StateChanged`.

I don't see any of those events in latest stable SignalR Core release 1.0.1.  I tried monitoring changes in `HubConnection.State` property, but that seems to mirror `HubConnection.Closed` event, and also takes 90 seconds to update.

What are the recommended workarounds to detect the change in connection faster?

**Comment by BrennanConroy** [comment]
I assume you're using LongPolling because 90 seconds is how long each poll takes. If you use WebSockets or SSE then the client will timeout in a more reasonable time ~30 seconds. If you want it to be faster you can change the keep alive and timeout values, see https://docs.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-2.1#configure-timeout-and-keep-alive-options

**Comment by Eternal21** [comment]
@BrennanConroy Thanks for the reply.  I was using default values, so your suggestion looked really good as the root cause.  Sadly changing the default LongPolling timeout from 90 seconds to 30 seconds via the following code had no effect.  It still took 90 seconds for the client to detect disconnection:

```
app.UseSignalR(routes =>
{
   routes.MapHub<PushHub>("/push", options =>
   {
      options.LongPolling.PollTimeout = TimeSpan.FromSeconds(30);
   });
});
```
The more I'm looking into the issue, the more I'm convinced it has something to do with how IIS recycles the .NET Core Service.  If I stop the service, SignalR detects disconnection immediately.  If I recycle the service it takes 90 seconds.  So my theory at this point is that SignalR connection truly exists for those 90 seconds, and is only terminated once the recycled app pool gets completely destroyed.

The problem is still the fact that the clients take 90 seconds to detect the fact that IIS is recycling the pool.  I tried setting `disallowOverlappingRotation` in IIS to true, to prevent old app pool from lingering, but it also had no effect.

EDIT: I finally stumbled upon the following setting in IIS app pool Advanced Settings: `Shut Down Time Limit`.  The default value is 90 seconds, and after changing it to 30 seconds, I finally got SignalR to disconnect after 30 seconds.  So the theory about old app pool being alive causing the problem seems to be correct.  In any case, I am closing this issue, since it's more of an IIS issue.
